src/NuGet.Services.AzureSearch/DatabaseAuxiliaryDataFetcher.cs

```diff
@@ -137,6 +152,52 @@ public async Task<HashSet<string>> GetVerifiedPackagesAsync()
                 }
             }
         }
+
+        public async Task<SortedDictionary<string, SortedSet<string>>> GetPackageIdToPopularityTransfersAsync()
+        {
+            var stopwatch = Stopwatch.StartNew();
+            var builder = new PackageIdToPopularityTransfersBuilder(_logger);
+            using (var connection = await _connectionFactory.OpenAsync())
+            using (var command = connection.CreateCommand())
+            {
+                command.CommandText = GetPopularityTransfersSql;
+                command.Parameters.Add(GetPopularityTransfersSkipParameter, SqlDbType.Int);
+                command.Parameters.AddWithValue(GetPopularityTransfersTakeParameter, GetPopularityTransfersPageSize);
+
+                // Load popularity transfers by paging through the database.
+                // We continue paging until we receive fewer results than the page size.
+                int currentPageResults;
+                int totalResults = 0;
+                do
+                {
+                    command.Parameters[GetPopularityTransfersSkipParameter].Value = totalResults;
+
+                    using (var reader = await command.ExecuteReaderAsync())
+                    {
+                        currentPageResults = 0;
+
+                        while (await reader.ReadAsync())
+                        {
+                            currentPageResults++;
+
+                            var fromId = reader.GetString(0);
+                            var toId = reader.GetString(1);
+
+                            builder.Add(fromId, toId);
+                        }
+                    }
+
+                    totalResults += currentPageResults;
+                }
+                while (currentPageResults == GetPopularityTransfersPageSize);
```

Comment:
In light of the recent bug, what do you think about switching to a >= 0 check instead? This mimic streams and makes fewer assumptions. Mainly a philosophical question.

Reply:
Writing this code definitely gave me the heebie jeebies 😅

For others that don't know the context: we recently had a bug on the Visual Studio Marketplace where we skipped extensions that we mitigated by tweaking some paging logic. The previous paging logic would fetch results until it found a batch that wasn't full. This was problematic as some batches weren't full due to extensions in invalid states (the extensions pointed to publishers that did not exist, causing these extensions to be skipped by a `JOIN`). We mitigated this by paging through results until we reach an empty batch (a >0 check).

To answer your question: the >0 check is not a perfect fix, it's a mitigation that sweeps the problem under the rug. The bug may reappear if there are more results in an invalid state than the batch size! In fact, using the >0 check may make it harder to detect that there are results in an invalid state. I would rather catch these kinds of issues quickly.